### PR TITLE
MTDSA-24890: shared code update with 'withResponseModifier' in Reques…

### PR DESCRIPTION
**Pull request by FergusHMRC**
…tHandler and optional + mandatory field validation in JsonErrorValidators

### CODE REVIEW

File: app/shared/controllers/RequestHandler.scala [hunk 1]
```diff
@@ -70,6 +70,9 @@ object RequestHandler {
     def withAuditing(auditHandler: AuditHandler): RequestHandlerBuilder[Input, Output] =
       copy(auditHandler = Some(auditHandler))
 
+    def withResponseModifier(modelHandler: Output => Output): RequestHandlerBuilder[Input, Output] =
```
Comment (mattstephens-hmrc) [inline app/shared/controllers/RequestHandler.scala:73]: could "modelHandler" be called "responseModifier" or sm?  ("modelHandler" doesn't really make sense)

File: app/shared/shared-code-changelog.md [hunk 2]
```diff
@@ -7,6 +7,10 @@ For the Shared Code update steps, see: https://confluence.tools.tax.service.gov.
 
 Place new items at the top, and auto-format the file...
 
+## July 9 2024: Updates from individual-calculations-api
+
+Minor additions to RequestHandler and JsonErrorValidators
```
Comment (mattstephens-hmrc) [inline app/shared/shared-code-changelog.md:12]: Maybe "Added RequestHandler.withResponseModifier"  & then sm similar for the JsonErrorValidators change on the next line?

File: test/shared/models/utils/JsonErrorValidators.scala [hunk 3]
```diff
@@ -94,6 +118,21 @@ trait JsonErrorValidators {
     }
   }
 
+  def testOptionalFields[A](json: JsValue)(property: String)(implicit rds: Reads[A]): Unit = {
+    s"the JSON is missing the optional property $property" should {
+
+      val jsonWithoutProperty = json.as[JsObject].-(property)
+
+      "not throw any errors" in {
+        jsonWithoutProperty.validate[A].isError shouldBe false
```
Comment (mattstephens-hmrc) [inline test/shared/models/utils/JsonErrorValidators.scala:127]: Can these be assigned to a "val result", and then "result shouldBe ...", more in line with the test pattern we generally use

**Review by LouiseWong98 — DISMISSED** [review]

File: app/shared/controllers/RequestHandler.scala [hunk 4]
```diff
@@ -160,15 +163,21 @@ object RequestHandler {
...
             EitherT[Future, ErrorWrapper, Result](Future.successful(Left(ErrorWrapper(ctx.correlationId, RuleRequestCannotBeFulfilledError))))
-          else
+          } else {
             for {
               parsedRequest   <- EitherT.fromEither[Future](validator.validateAndWrapResult())
               serviceResponse <- EitherT(service(parsedRequest))
             } yield doWithContext(ctx.withCorrelationId(serviceResponse.correlationId)) { implicit ctx: RequestContext =>
-              handleSuccess(parsedRequest, serviceResponse)
+              responseModifier match {
+                case Some(responseHandler) =>
```
Comment (mattstephens-hmrc) [inline app/shared/controllers/RequestHandler.scala:174]: could this just be "Some(modifier)"?  "responseHandler" is a bit confusing here alongside the RequestHandler etc

File: app/shared/shared-code-changelog.md [hunk 5]
```diff
@@ -7,10 +7,22 @@ For the Shared Code update steps, see: https://confluence.tools.tax.service.gov.
 
 Place new items at the top, and auto-format the file...
 
+
+## July 9 2024: Updates from individual-calculations-api
```
Comment (mattstephens-hmrc) [inline app/shared/shared-code-changelog.md:11]: If you're making the above change then could also change to July 10 while you're there

**Review by LouiseWong98 — APPROVED** [review]

**Pull request merged**